### ISSUE TITLE
"Objects changed outside of Terraform" shown wrongly for various resources

### ISSUE THREAD

**Issue by dimisjim** [hashicorp/terraform-provider-aws#19727]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

## Please open a new issue

1. Since this issue was opened on Jun 9, 2021, we have had two major rounds of fixes (#21968, #23288 and #28836 et al.) that seem to have _improved_ and/or eliminated changed reports being shown wrongly
2. In order to alert maintainers and the community to what is _still_ not working, open a [new issue](https://github.com/hashicorp/terraform-provider-aws/issues/new/choose) with a title similar to `[Bug]: ec2/instance objects have changed outside of Terraform`. _Check to make sure one doesn't already exist for the resource you are working on._

### Terraform CLI and Terraform AWS Provider Version

```
terraform version: 1.0.0
aws provider version: 3.44.0
```

### Affected Resource(s)

Related to jsonencode list items ordering
* aws_s3_bucket_policy
* aws_iam_role

Related to `latest_restorable_time` attribute with drifts
* aws_db_instance

and more

### Expected Behavior

No change was made outside of terraform so these should not be visible there as well

### Actual Behavior

Terraform thinks that something was changed outside terraform (which is not the case) 

### References

This issue was first submitted in https://github.com/hashicorp/terraform/issues/28911

**Comment by pm-proctors** [comment]
Also affects `aws_wafv2_web_acl` resources; it is unclear why, as the attributes list shows no changes. Applying the refresh plan has no effect — it will still be marked as changed in the next plan.

Observed on provider version `v3.45.0`.

**Comment by nigel-codaio** [comment]
Also affects `aws_kms_key` resources - the policy statement principals are changing order:
```
  ~ resource "aws_kms_key" "key" {
        id                       = ...
      ~ policy                   = jsonencode(
          ~ {
              ~ Statement = [
                  ~ {
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::XXXXXXXXXXXX:root",
                                "arn:aws:iam::YYYYYYYYYYYY:root",
                              + "arn:aws:iam::XXXXXXXXXXXX:root",
                            ]
                        }
                        # (4 unchanged elements hidden)
                    },
...
```

**Comment by pikesley** [comment]
> Also affects `aws_kms_key` resources - the policy statement principals are changing order:
> 
> ```
>   ~ resource "aws_kms_key" "key" {
>         id                       = ...
>       ~ policy                   = jsonencode(
>           ~ {
>               ~ Statement = [
>                   ~ {
>                       ~ Principal = {
>                           ~ AWS = [
>                               - "arn:aws:iam::XXXXXXXXXXXX:root",
>                                 "arn:aws:iam::YYYYYYYYYYYY:root",
>                               + "arn:aws:iam::XXXXXXXXXXXX:root",
>                             ]
>                         }
>                         # (4 unchanged elements hidden)
>                     },
> ...
> ```

Yes, we got hit with exactly this :/

**Comment by snstanton** [comment]
This seems to affect anything with a policy.  The principal list appears to come back in a non-deterministic order.

**Comment by tobypinder** [comment]
> This seems to affect anything with a policy. The principal list appears to come back in a non-deterministic order.

Confirming this with an additional example, `aws_iam_role` resources are affected via the `assume_role_policy`, also as part of the Principal block non-determinism. Alphabetizing from the start did not appear to mitigate the issue for us.

**Comment by jbscare** [comment]
This also affects things that change outside of Terraform, but which you've told Terraform to ignore.

Two places we've seen that are (1) tags in a lifecycle ignore_changes list; and (2) when an ELB isn't configured in Terraform to have listeners, because we have separate orchestration tools that add and remove them. In both of those cases, Terraform used to not show changes when those things changed behind the scenes, but now it does.

I'm not sure if that's a provider side thing that fits with this ticket, or if the ignored-changes problem belongs with https://github.com/hashicorp/terraform/issues/28803.

**Comment by Dunedan** [comment]
The `aws_backup_vault` resource is affected as well:
```
Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # aws_backup_vault.test has been changed
  ~ resource "aws_backup_vault" "custom_idp" {
        id              = "test"
        name            = "test"
      ~ recovery_points = 9 -> 10
        # (3 unchanged attributes hidden)
    }
```

**Comment by danopia** [comment]
I'm seeing an `aws_iam_role` policy as changed without even having a list being reordered, or any other change visibly pointed out. Terraform apparently includes `# whitespace changes` in the output when this happens.

```hcl
Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # module.identity.aws_iam_role.role has been changed
  ~ resource "aws_iam_role" "role" {
      ~ assume_role_policy    = jsonencode( # whitespace changes
            {
                Statement = [
                    {
                        # [snip for brevity]
                    },
                    {
                        # [snip for brevity]
                    },
                ]
                Version   = "2012-10-17"
            }
        )
        # [snip for brevity]
    }
```

**Comment by shouze** [comment]
I confirm it's still the case with `aws_wafv2_web_acl` (provider version `3.69.0`).

Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

```hcl
# aws_wafv2_web_acl.gogaille_cloudfront has changed
~ resource "aws_wafv2_web_acl" "some_resource" {
      id          = "731f3d3e-b5b9-48fe-9f91-3b73b7230839"
      name        = "some-name"
      tags        = {}
      # (6 unchanged attributes hidden)



      # (8 unchanged blocks hidden)
  }
```

**Comment by eschulma** [comment]
I also see this with` aws_wafv2_web_acl` (provider version `3.72.0`). My output is almost identical to what @shouze posted above. Even deleting and recreating the ACL doesn't help. I do use `AWSManagedRulesAmazonIpReputationList` as one of my rules. Here is the full config:

```
resource "aws_wafv2_web_acl" "cloudfront" {
  name        = "acl-sentrylink-cloudfront"
  description = "Our website CloudFront ACL."
  scope       = "CLOUDFRONT"
  tags        = {}

  default_action {
    allow {}
  }

  rule {
    name     = "trusted-ip-access"
    priority = 0

    action {
      allow {}
    }

    statement {
      ip_set_reference_statement {
        arn = aws_wafv2_ip_set.cloudfront_trusted_ips.arn
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = false
      metric_name                = "trusted-ip-access"
      sampled_requests_enabled   = true
    }
  }

  rule {
    name     = "block-blacklisted-ips"
    priority = 1

    action {
      block {}
    }

    statement {
      ip_set_reference_statement {
        arn = aws_wafv2_ip_set.cloudfront_blacklist_ips.arn
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = true
      metric_name                = "block-blacklisted-ips"
      sampled_requests_enabled   = true
    }
  }

  rule {
    name     = "general-cf-rate-limit"
    priority = 2

    action {
      block {}
    }

    statement {
      rate_based_statement {
        limit              = 2500
        aggregate_key_type = "IP"
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = true
      metric_name                = "general-cf-rate-limit"
      sampled_requests_enabled   = true
    }
  }

  rule {
    name     = "AWS-AWSManagedRulesAmazonIpReputationList"
    priority = 3

    override_action {
      none {}
    }

    statement {
      managed_rule_group_statement {
        name        = "AWSManagedRulesAmazonIpReputationList"
        vendor_name = "AWS"
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = true
      metric_name                = "AWS-AWSManagedRulesAmazonIpReputationList"
      sampled_requests_enabled   = true
    }
  }

  visibility_config {
    cloudwatch_metrics_enabled = true
    metric_name                = "acl-sentrylink-cloudfront"
    sampled_requests_enabled   = true
  }
}
```

**Comment by byF** [comment]
Is there any plan to do something about this? 🤔

**Comment by gabrielsyapse** [comment]
For reference, the original issue left open since June 2021 https://github.com/hashicorp/terraform/issues/28803

**Comment by esteban1983cl** [comment]
Same issue here

**Comment by YakDriver** [comment]
At least some, hopefully many, of these issues should have been resolved with #21968 (v3.70.0). Letting us know exactly what is still occurring post v3.70.0 (such as [this above](https://github.com/hashicorp/terraform-provider-aws/issues/19727#issuecomment-1017952112)) is very helpful. For bonus points, you can create a new issue for specific reasons (not necessarily specific resources though) and refer back to this one.

**Comment by clint-truss** [comment]
> I confirm it's still the case with `aws_wafv2_web_acl` (provider version `3.69.0`).
> 
> Terraform detected the following changes made outside of Terraform since the last "terraform apply":
> 
> ```terraform
> # aws_wafv2_web_acl.gogaille_cloudfront has changed
> ~ resource "aws_wafv2_web_acl" "some_resource" {
>       id          = "731f3d3e-b5b9-48fe-9f91-3b73b7230839"
>       name        = "some-name"
>       tags        = {}
>       # (6 unchanged attributes hidden)
> 
> 
> 
>       # (8 unchanged blocks hidden)
>   }
> ```

I am seeing this issue with aws_wafv2_web_acl with AWS provider 3.74.0 and terraform 1.1.5.

**Comment by trallnag** [comment]
In my case it does not make a difference if I am using an inline policy or a dedicated policy / attachment resource.

```
Terraform v1.1.5
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.74.3
+ provider registry.terraform.io/hashicorp/external v2.2.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
```

The policy JSON can be found here:

https://raw.githubusercontent.com/kubernetes-sigs/aws-load-balancer-controller/v2.4.0/docs/install/iam_policy.json

**Comment by esteban1983cl** [comment]
Hi everyone, is there some news about this ?
Thanks in advance ?

**Comment by nantiferov** [comment]
According to this comment, since v1.2 (not released yet) it should became kinda better
https://github.com/hashicorp/terraform/issues/28803#issuecomment-1072740861

> Starting with v1.2, the goal for the refresh report is that only external changes which may have contributed to changes in the plan will be shown. This means in most cases, unused attributes changing outside of terraform will not show up in the normal plan output. If there are no changes in the plan, no external changes will be shown in the CLI at all. All refresh information is still stored within the plan, and if a user wants to see all external changes of resources in the CLI, a refresh-only plan can be used.

P.S. Btw, sed could be used to cut off this nice feature (macOS example)
```shell
terraform plan|sed -n '/Objects have changed outside of Terraform/,/─────────────────────────────────────────────────────────────────────────────/!p'
```

**Comment by nantiferov** [comment]
And v1.2 released, but this stuff still there 😢 (tested with latest TF 1.2.1 and aws provider 4.15.1)
One year later after this "great" feature was implemented we're still there 🤦 

Every plan/apply run I'm getting diff in IAM policies for example. And it's terrible with relatively big states.

Like in this case just order changes
```terraform
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::1111:root",
                                "arn:aws:iam::6666:role/some-role",
                              + "arn:aws:iam::1111:root",
                            ]
                        }
```

And next time
```terraform
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam:: 6666:role/some-role",
                                "arn:aws:iam::1111:root",
                              + "arn:aws:iam:: 6666:role/Test-Node-Server",
                            ]
                        }
```

I ended up now just cutting this from output with different wrappers.

**Comment by nantiferov** [comment]
Tested with TF v1.2.1/4.16.0, the same.

It's `aws_sqs_queue`, `policy` property in my case, I guess it depends on resource. Some fixed, some - not.

```hcl
resource "aws_sqs_queue" "this" {
      ~ policy                            = jsonencode(
           {
...
                      ~ Principal = {
                          ~ AWS = [
                              - "arn:aws:iam::6666:role/some-role",
                                "arn:aws:iam::1111:root",
                              + "arn:aws:iam::6666:role/some-role",
                            ]
                        }
...
          }
```

**Comment by gilad9366** [comment]
Tested with TF v1.3.6/v4.46.0, getting this "false" change on each apply:
```
resource "aws_autoscaling_group" "test" {
        id                  = "xxx"
      + load_balancers      = []
        name                = "yyy"
      + suspended_processes = []
        # (22 unchanged attributes hidden)
        # (11 unchanged blocks hidden)
    }
```

**Comment by YakDriver** [comment]
@gilad9366 Please open an issue for `aws_autoscaling_group` if this is still an issue so that we can prioritize and get eyes on that specific problem.

@nantiferov I would love to hear back from you if you are still facing problems with drift using `v4.50.0`. Please see #23288.

**Comment by YakDriver** [comment]
## Please open a new issue

1. Since this issue was opened on Jun 9, 2021, we have had two major rounds of fixes (#21968, #23288 and #28836 et al.) that seem to have _improved_ and/or eliminated changed reports being shown wrongly
2. In order to alert maintainers and the community to what is _still_ not working, open a [new issue](https://github.com/hashicorp/terraform-provider-aws/issues/new/choose) with a title similar to `[Bug]: ec2/instance objects have changed outside of Terraform`. _Check to make sure one doesn't already exist for the resource you are working on._
